### ISSUE TITLE
Using multiple file browsers on a single node form breaks functionality

### ISSUE THREAD

**Issue by bobchristenson** [backdrop/backdrop-issues#6924]
## Description of the bug

If you have a node form in which you have multiple methods of using a file browser (ex: a filefield with 'select existing file' plus a wysiwyg 'image' button where you 'select from library'), after using the first instance successfully, the second instance has broken functionality. 

I think there are multiple combinations of functionality which can cause this, but in my case I found it while using an image file field while paired with the CKEditor 5 image upload button within a WYSIWYG.  (using other wysiwyg options is reported to cause this behavior as well)

## Steps To Reproduce

To reproduce the behavior:
1. Create a content type with an image file field which uses the 'select existing file 'functionality so you can select an existing file from the media library using the library modal selector
2. Setup CKeditor 5 to use an image button and allow re-use of existing files
3. Create a new node using this type and use the 'select existing file' on the image filefield.  Choose a file and this will load the image into the field properly
4. Now go use the WYSIWYG image button in your body (textarea).  Click the image button, choose the tab to 'select from library'. Now use the search function to find an existing image by filename.

## Actual behavior
When you click apply after searching for the image by filename, the apply button is broken and redirects you to a strange url which includes the filename you searched for.  Something like: http://mysite.com/?filename=file_that_was_searched.jpg

## Expected behavior
Obviously, the apply button should, using javascript, filter the library image list by your search and allow you to choose the image.

## Additional information

My guess is that using the first library modal to choose an image is then breaking any additional uses of the modal on the page, possibly due to scope issues.

Note that you can do different combinations and get similar behavior.  I've confirmed it works in reverse (do the WYSIWYG image first (succesfully) and then you'll see that the file field image library browser is broken.

**Comment by indigoxela** [comment]
I didn't dig too deep, but it seems like the image field AJAX and the editor dialog AJAX badly interfere.

Loading the form part after submissions (by the other variant) seems to add `+1` to the input ID, and the other variant loses its event listener (caution: only a working thesis 😜).

What I mean is this: `<input type="text" id="edit-filename--2" name="filename" value="" size="30" maxlength="128" class="form-text">` AJAX causes this `edit-filename--2`, `edit-filename--3`.

Note that this adding up with the ID is no problem, if sticking with one AJAX variant - either file field *or* editor dialog, but switching between them breaks things.

Pretty annoying, if one has an editor open with lots of changes... and then one gets "thrown" to the base URL.

**Comment by indigoxela** [comment]
Yeah, that was quite some riddle - as expected. Turned out - the problem's caused by Views AJAX.

Here's now a PR that fixes it (hopefully with no side effects): https://github.com/backdrop/backdrop/pull/5083

The failing test is unrelated (spellcheck) and can't get fixed, anyway.

Background: the info about the exposed form was outdated, when switching between file and editor library loading. Making the jQuery selector more specific (unless it's an exposed form in a block), grabs the right data and adding the event listeners to the "apply" button works properly. There's now also a proper jQuery.once identifier.

**Comment by indigoxela** [comment]
Easy testing can be done with node/add/post, as posts have an editor (CKE) AND an image field.

Testing beyond that: create an AJAx view with an exposed form in a block. Test all kinds of views with exposed form already available, like admin/content ... whatever comes to your mind.

Steps for the "easy test":

- Go to node/add/post
- Click in the editor, click on the image toolbar button to open the dialog, switch over to "Select from library" by clicking the link (the image library view with its thumbnails loads then)
- Insert a search term into the Filename input and submit (no need to insert an image...)
- Click on the "Select existing file" button in the Image field below
- Insert a search term into the Filename input and submit
- You shouldn't end up on the site's start page 😉 

Also do it the other way round: start with the Image field and after that open the editor dialog and try to use a search term there.

**Comment by argiepiano** [comment]
@indigoxela thanks for the steps, but I can't follow what you are saying. I'm trying the second step above. I'm not clear what you mean by "open the image library" and then "switch over to Select from library".

Can you please edit/clarify/expand those steps?

**Comment by indigoxela** [comment]
> Can you please edit/clarify/expand those steps?

I tried - is it clearer now?

**Comment by argiepiano** [comment]
Yes, I'm now able to reproduce. I'll test the PR next.

**Comment by yorkshire-pudding** [comment]
Tested and was able to reproduce the issue - I've seen this on a client site and wondered what it was. 
Applied the patch and the issue was fixed.
Brilliant work @indigoxela  - thank you.

**Comment by argiepiano** [comment]
Works for me too (@yorkshire-pudding beat me changing the labels). I've looked at the code, and it makes sense, but I do have a question: why is the behavior of an exposed views form in a block different from the "regular" views form? I see there is an if statement to accommodate this. Wouldn't this fix work the same in both situations?

**Comment by indigoxela** [comment]
Awesome, many thanks to everyone testing this PR. 🙏 

> question: why is the behavior of an exposed views form in a block different from the "regular" views form?

If the exposed form is part of the view, it gets replaced when reloading the view. This doesn't happen with exposed forms in blocks, which are outside the view.

In exposed forms in blocks the selector based on the view's dom_id wouldn't work, as it's _outside_ the view. But it's also not necessary to update it, as it's not changed by reloading the view content.

What happens when different scripts trigger a views update and the exposed form is part of the view is, that the id of the form item changes (edit-filename turns into edit-filename--2...) and the event listener on the button also gets wiped out. The identifier for the views exposed form stays the same, though, and the jQuery.once() has already run on it.
The surrounding views dom_id value does change, so using that one makes sure, the "once" runs again when necessary. We do need this "once", as otherwise the listener gets attached multiple times (ajax...).

**Comment by argiepiano** [comment]
Thanks for the explanation, @indigoxela. I get it now. Code looks good to me, and I think this is RTBC.

**Comment by quicksketch** [comment]
Nice work @indigoxela! Thank you @argiepiano and @yorkshire-pudding for testing the fix; and thanks to @bobchristenson for filing this issue with such a good set of steps to reproduce! Merged https://github.com/backdrop/backdrop/pull/5083 into 1.x and 1.31.x.
